File: 38_sliding_windows/38_14_shortest_period_with_over_20_sales.py
import math

def shortest_period_with_over_20_sales(sales: list[int])->int:
  l, r = 0, 0
  cur_min = math.inf

  while r < len(sales):
    window = sum(sales[l:r+1])
    if window > 20 and r - l + 1 < cur_min:
      cur_min = r - l + 1
      l += 1
    else:
      r += 1

  return cur_min if cur_min != math.inf else -1
  

if __name__ == "__main__":
  # Example1
  sales = [5, 10, 15, 5, 10]
  print(shortest_period_with_over_20_sales(sales))

  # Example2
  sales = [5, 10, 4, 5, 10]
  print(shortest_period_with_over_20_sales(sales))

  # Example3
  sales = [5, 5, 5, 5]
  print(shortest_period_with_over_20_sales(sales))

File: 38_sliding_windows/38_14_shortest_period_with_over_20_sales.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <numeric>

int shortest_period_with_over_20_sales(std::vector<int> sales) {
  int l = 0, r = 0;
  int cur_min = INT_MAX;

  while (r < sales.size()) {
    int window = std::accumulate(sales.begin() + l, sales.begin() + r + 1, 0);
    if (window > 20 && r - l + 1 < cur_min) {
      cur_min = r - l + 1;
      l += 1;
    } else {
      r += 1;
    }
  }

  return cur_min != INT_MAX ? cur_min : -1;
}

int main() {
  // Example1
  std::vector<int> sales1 = {5, 10, 15, 5, 10};
  std::cout << shortest_period_with_over_20_sales(sales1) << std::endl;

  // Example2
  std::vector<int> sales2 = {5, 10, 4, 5, 10};
  std::cout << shortest_period_with_over_20_sales(sales2) << std::endl;

  // Example3
  std::vector<int> sales3 = {5, 5, 5, 5};
  std::cout << shortest_period_with_over_20_sales(sales3) << std::endl;

  return 0;
}
